content/testing.md

```diff
@@ -1,23 +1,158 @@
+
+### Internal Unit Tests
+
+In most cases, packages can be tested using external tests only. When writing tests for a package called `xyz`, the external tests should be in the same folder, but in a package called `xyz_test`. This case is [handled by Go natively](https://pkg.go.dev/cmd/go@master#hdr-Test_packages) and will therefore not result in complaints about there being two different packages within the same directory. Of course, there are exceptions. If you need to test some internal logic, those tests must be in a file suffixed with `_internal_test.go`.
+
+### Mocks
+
+When it comes to mocking dependencies for tests, we prefer to use simple hand-made mocks rather than to use testing frameworks to generate them. The Go language makes it easy to do so elegantly by creating structures that implement the interfaces for dependencies of the tested code and exposing functions that match the interface's signature as attributes that can be overridden externally.
+
+```go
+{!mock_example.md!}
+```
+
+Using those mocks is as simple as instantiating a baseline version of the mock and setting its attributes to the desired functions:
+
+```go
+{!mock_test_example.md!}
+```
+
+### Pseudorandom Generic Values
+
+When using test data for unit tests, it is always a good idea to use random generated data as the inputs. This avoids the bias where a test passes because it is given a valid set of inputs while some other inputs might have highlighted a flaw in the logic, by using an _unconstrained_ data set.
+
+In order for the tests to be repeatable and for results to be consistent though, the given inputs should not be completely random, but instead they should be pseudorandom, with the same initial seed, to ensure the same sequence of "random" tests.
+
+Here is an example of such a value being generated.
+
+```go
+{!generic.md!}
+```
+
+!!! warning
+
+    While randomly generating valid inputs makes sense, randomly generating invalid inputs does not. In the case of invalid inputs, it is much better to have an exhaustive list of all types of cases that are expected to be invalid and always test each one of them.
+
+### Parallelization
+
+Since the version `1.7` of Go, tests can be run in parallel. This can be done by calling [`t.Parallel`](https://pkg.go.dev/testing#T.Parallel) in each subtest. Calling this function signals that the test is to be run in parallel with (and only with) other parallel tests, and the amount of tests running in parallel is limited by the value of [`runtime.GOMAXPROCS`](https://pkg.go.dev/runtime#GOMAXPROCS).
+
+There are multiple advantages to parallelizing tests:
+
+* It ensures that regardless of the order in which inputs are given, components behave as expected.
+* It maximizes performance, which in turns results in a faster CI and a faster workflow for everyone.
+* By making tests run faster, it allows you to write more tests and therefore produce more actionable data to find bugs as well as improve tests cases and coverage.
```

Comment:
I don't think this bullet point adds much value tbh, it kind of falls in the "faster workflow" category of benefits.

Reply:
There is the advantage of random ordering and concurrent execution upon the same component though. I agree that two points are about the speed, but the two others are about completely different subjects, are they not? 😄

Reply:
No no, I agree completely, I only thought that the third bullet point is a bit of an elaboration of the second one.. But all the mentioned things are definitely relevant and should be mentioned I think :)